### DISCUSSION TITLE
Оформление slicing на fsh

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#227] [General]
*

**Comment by GorbachevaT** [comment]
Добрый день, коллеги!
Прошу помощи.
Пробую переписать на FSH профиль Организация
![image](https://user-images.githubusercontent.com/127845097/236407073-9f9fa589-ef6f-4350-bf3b-dcb49c343321.png)

Есть вопросы:
1.	Не понятно правило ^slicing.rules = #open или #closed. В каких случаях может быть полезно и применяется #open? Это ведь, если я правильно поняла, разрешение передавать больше чем описано?
2.	Возможно ли и как оформить  ДО слайсинга параметра identifier, его поле assigner, которое в нарезке не участвует?
3.	Ну и основной вопрос – как правильно оформить слайсинг индентификатора Организации по типам реквизитов организации - ИНН, ФРМО, ОГРН?

**Comment by GorbachevaT** [comment]
Здесь текст на fsh, то, что проходит валидацию. Дальше как ни крутила, все ошибка. Как это правильно написать в нарезке по url? Подозреваю, что я не правильно подхожу, может там другая логика нарезки вообще...
```
Profile:     CoreOrganization
Id:          core-organization
Parent:      Organization
Title:       "Organization"
Description: "Профиль Organisation описывает субъекты хозяйственной деятельности в области здравоохранения. Определены специфические идентификаторы - ОГРН, ФРМО, ИНН."


* name 1..1 MS
* name ^short = "Полное название организации"
* name only string

* alias 0..*
* alias ^short = "Сокращенное наименование организации"
* alias only string


* identifier 0..* 

//* assigner 0..1 




* identifier.extension ^slicing.discriminator.type = #type
* identifier.extension ^slicing.discriminator.path = "url" 
* identifier ^slicing.rules = #open
* identifier ^slicing.description = "Нарезка по типам реквизитов организации - ИНН, ФРМО, ОГРН"

* identifier contains 
  ИНН  0..1 and
  ФРМО 0..1 and
  ОГРН 0..1 
  
/*
* identifier[ИНН] contains
  system 1..1
  
* identifier[ФРМО] contains
  system 1..1  
  
* identifier[ОГРН] contains
  system 1..1
*/



* address 0..*
* address ^short = "Адрес организации"
* address only Address //CoreAddress = http://http://fhir-ru.zendoc.me/core.datatypes.Address
```

**Comment by Vrodvi** [comment]
Разбор по вопросам:

> 1. Не понятно правило ^slicing.rules = #open или #closed. В каких случаях может быть полезно и применяется #open? Это ведь, если я правильно поняла, разрешение передавать больше чем описано?
- #open - обозначение, что, помимо указанных срезов, в атрибуте могут быть добавлены дополнительные срезы
- #closed - обозначение, что, помимо указанных срезов, в атрибуте уже нельзя добавлять новые срезы
- Желательно по-умолчанию использовать открытый режим (#open) нарезки, тем более в корневых профилях (Ru Core), чтобы разрешить последующее профилирование.

> 2. Возможно ли и как оформить ДО слайсинга параметра identifier, его поле assigner, которое в нарезке не участвует?
- Да, в профиле можно накладывать общие ограничения на атрибут, а затем ещё и добавить нарезку этого же атрибута с дополнительными специфическими ограничениями
- Общее ограничение:
```
* identifier.assigner 0..1
* identifier.assigner only Reference(CoreOrganization)
```
- специфические ограничения в срезах
```
* identifier[ИНН] ^short = "Идентификационный номер налогоплательщика — цифровой код, упорядочивающий учёт налогоплательщиков в Российской Федерации. Присваивается налоговой записи как юридических, так и физических лиц Федеральной налоговой службой России."
* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
```

> 3. Ну и основной вопрос – как правильно оформить слайсинг индентификатора Организации по типам реквизитов организации - ИНН, ФРМО, ОГРН?
- Правила оформления срезов (слайсинга):  
  - описание - https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#contains-rules-for-slicing
  - пример - https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#add-element-rules:~:text=Complete%20Slicing%20Example
  - параметры срезов - описание того, каким образом можно отличить один срез от другого - описано тут http://hl7.org/fhir/R4/profiling.html#discriminator
- Указанные фрагменты описания желательно перевести на русский язык и приложить в общую документацию Ru Core, в методические материалы для упрощения последующей работы
- Пример описания срезов на примере определения реквизитов ИНН, ФРМО, ОГРН для организации:
```
* identifier ^slicing.discriminator.type = #value
* identifier ^slicing.discriminator.path = "system" 
* identifier ^slicing.rules = #open
* identifier ^slicing.description = "Нарезка по типам реквизитов организации - ИНН, ФРМО, ОГРН"

* identifier contains 
  ИНН  0..1 and
  ФРМО 0..1 and
  ОГРН 0..1 
  
* identifier[ИНН] ^short = "Идентификационный номер налогоплательщика — цифровой код, упорядочивающий учёт налогоплательщиков в Российской Федерации. Присваивается налоговой записи как юридических, так и физических лиц Федеральной налоговой службой России."
* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
  
* identifier[ФРМО] ^short = "Уникальный идентификатор МО в справочнике ФРМО"
* identifier[ФРМО].system = "http://fhir.ru/core/systems/frmo"  
  
* identifier[ОГРН] ^short = "Основной государственный регистрационный номер юридического лица (индивидуального предпринимателя)"
* identifier[ОГРН].system = "http://fhir.ru/core/systems/ogrn"
```

----
Разберём проблемные места в представленном описании профиля:

```
* name only string
* alias only string
```
- если профиль не модифицирует стандартные типы данных, то их не нужно указывать в профиле 
----
```
//* assigner 0..1 
```
- это вложенный атрибут и его ограничения нужно описывать через указанием родительского узла через точку:
```
* identifier.assigner 0..1
* identifier.assigner only Reference(CoreOrganization)
```
----

```
* identifier.extension ^slicing.discriminator.type = #type
* identifier.extension ^slicing.discriminator.path = "url" 
* identifier ^slicing.rules = #open
* identifier ^slicing.description = "Нарезка по типам реквизитов организации - ИНН, ФРМО, ОГРН"
```
- тут ошибочно указано, что срез применяется к "расширению" - "identifier.extension". Срез делается для атрибута identifier, поэтому правильно должно быть так:
```
* identifier ^slicing.discriminator.type = #xxx
* identifier ^slicing.discriminator.path = "xxx" 
```
- в ^slicing.discriminator.type указывается критерий по которому можно разделить один срез от другого. Все типы описаны тут - http://hl7.org/fhir/R4/profiling.html#discriminator. В данном месте разделение срезов по типу ресурсов ( #type) не подходит. Правильный вариант будет разделение срезов по значению (#value), а конкретно по значению "пространства идентификаторов", которое указывается в атрибуте "system" (identifier.system), что указывается через нотацию ^slicing.discriminator.path = "XXX", где XXX - нужно установить в путь до атрибута, значение которого нужно проверять для разведения срезов (в нашем случае - system).
- итоговый фрагмент для описания срезов:
```
* identifier ^slicing.discriminator.type = #value
* identifier ^slicing.discriminator.path = "system" 
* identifier ^slicing.rules = #open
* identifier ^slicing.description = "Нарезка по типам реквизитов организации - ИНН, ФРМО, ОГРН"
```
----
```
/*
* identifier[ИНН] contains
  system 1..1
  
* identifier[ФРМО] contains
  system 1..1  
  
* identifier[ОГРН] contains
  system 1..1
*/
```
- описание ограничений на атрибуты внутри среза нужно делать не через "contains", а через нотацию через точку - указание родительского узла с уточнением среза в квадратных скобках и далее через точку указывать нужный атрибут для огарничения, т.е. в нашем случае должно быть:
```
* identifier[ИНН] ^short = "Идентификационный номер налогоплательщика — цифровой код, упорядочивающий учёт налогоплательщиков в Российской Федерации. Присваивается налоговой записи как юридических, так и физических лиц Федеральной налоговой службой России."
* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
  
* identifier[ФРМО] ^short = "Уникальный идентификатор МО в справочнике ФРМО"
* identifier[ФРМО].system = "http://fhir.ru/core/systems/frmo"  
  
* identifier[ОГРН] ^short = "Основной государственный регистрационный номер юридического лица (индивидуального предпринимателя)"
* identifier[ОГРН].system = "http://fhir.ru/core/systems/ogrn"
```
----

```
* address 0..*
* address ^short = "Адрес организации"
* address only Address //CoreAddress = http://http://fhir-ru.zendoc.me/core.datatypes.Address
```
- для того чтобы переопределить тип данных на ограниченный тип данных нужно предварительно описать "профиль" на этот ограниченный тип данных, т.е. в нашем случае:
```
Profile:     CoreAddress
Id:          core-address
Parent:      Address
Title:       "Адресс (Ru Core)"
Description: "Адрес с кодом адреса и кодом дома по ФИАС"

// TODO - нужно описать профиль на тип данных с предварительным описанием двух расширений, которые в нём используются
```

**Comment by Vrodvi** [comment]
Полный пример:
```
Profile:     CoreAddress
Id:          core-address
Parent:      Address
Title:       "Адресс (Ru Core)"
Description: "Адрес с кодом адреса и кодом дома по ФИАС"

// TODO - нужно описать профиль на тип данных с предварительным описанием двух расширений, которые в нём используются


Profile:     CoreOrganization
Id:          core-organization
Parent:      Organization
Title:       "Organization"
Description: "Профиль Organisation описывает субъекты хозяйственной деятельности в области здравоохранения. Определены специфические идентификаторы - ОГРН, ФРМО, ИНН."

* name 1..1 MS
* name ^short = "Полное название организации"

* alias 0..*
* alias ^short = "Сокращенное наименование организации"

* identifier 0..* 

* identifier.assigner 0..1
* identifier.assigner only Reference(CoreOrganization)

* identifier ^slicing.discriminator.type = #value
* identifier ^slicing.discriminator.path = "system" 
* identifier ^slicing.rules = #open
* identifier ^slicing.description = "Нарезка по типам реквизитов организации - ИНН, ФРМО, ОГРН"

* identifier contains 
  ИНН  0..1 and
  ФРМО 0..1 and
  ОГРН 0..1 
  
* identifier[ИНН] ^short = "Идентификационный номер налогоплательщика — цифровой код, упорядочивающий учёт налогоплательщиков в Российской Федерации. Присваивается налоговой записи как юридических, так и физических лиц Федеральной налоговой службой России."
* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
  
* identifier[ФРМО] ^short = "Уникальный идентификатор МО в справочнике ФРМО"
* identifier[ФРМО].system = "http://fhir.ru/core/systems/frmo"  
  
* identifier[ОГРН] ^short = "Основной государственный регистрационный номер юридического лица (индивидуального предпринимателя)"
* identifier[ОГРН].system = "http://fhir.ru/core/systems/ogrn"

* address 0..*
* address ^short = "Адрес организации"
* address only CoreAddress


Instance:   core-organization-instance-1
InstanceOf: CoreOrganization
Usage: #example

* name = "Полное название организации"

* alias = "Сокращенное наименование организации"

* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
* identifier[ИНН].value = "номер ИНН"
* identifier[ИНН].assigner = Reference(core-organization-instance-2)
  
* identifier[ФРМО].system = "http://fhir.ru/core/systems/frmo"  
* identifier[ФРМО].value = "идентификатор ФРМО"  
* identifier[ФРМО].assigner = Reference(ore-organization-instance-2)
  
* identifier[ОГРН].system = "http://fhir.ru/core/systems/ogrn"
* identifier[ОГРН].value = "номер ОГРН"
* identifier[ОГРН].assigner = Reference(core-organization-instance-2)

* address.text = "Адрес"


Instance:   core-organization-instance-2
InstanceOf: CoreOrganization
Usage: #example

* name = "Полное название организации 2"

* alias = "Сокращенное наименование организации 2"

* identifier[ИНН].system = "http://fhir.ru/core/systems/inn"
* identifier[ИНН].value = "номер ИНН 2"
  
* identifier[ФРМО].system = "http://fhir.ru/core/systems/frmo"  
* identifier[ФРМО].value = "идентификатор ФРМО 2"  
  
* identifier[ОГРН].system = "http://fhir.ru/core/systems/ogrn"
* identifier[ОГРН].value = "номер ОГРН 2"

* address.text = "Адрес 2"
```
![image](https://user-images.githubusercontent.com/92928137/236483975-bc17ea4e-0493-4960-bd16-c99ea5fb7618.png)

**Comment by GorbachevaT** [comment]
Большое спасибо! Очень познавательно и доходчиво!

**Comment by Vrodvi** [comment]
Замечания по странице http://fhir.ru/support.fsh-development.slicing
(в связи с запросом на http://fhir.ru/meetings.2023-05-16)

- немного сбито форматирование
  - ссылки не отображаются как ссылки
  - звёздочка не понятно что обозначает  
- на странице не указана ссылка на эту дискуссию, так что не понятно куда писать замечания по наполнению
(на других страницах иногда указывается, например:
![image](https://github.com/fhir-ru/core/assets/92928137/33e28ce2-793a-4511-85e8-b1dec61aef9b))

----  
  
![image](https://github.com/fhir-ru/core/assets/92928137/8c8b6cc9-d1d5-4bb7-abfb-c9cb5e08de16)
